File: binaryTreeInorderTraversal.py
# problem: https://leetcode.com/problems/binary-tree-inorder-traversal/
# Runtime: 32 ms, faster than 71.02% of Python3 online submissions for Binary Tree Inorder Traversal.
# Memory Usage: 14.3 MB, less than 13.33% of Python3 online submissions for Binary Tree Inorder Traversal.


from typing import List

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        answer = []

        def traverse(node):
            if not node:
                return
            traverse(node.left)
            answer.append(node.val)
            traverse(node.right)
        traverse(root)
        return answer

File: binaryTreeInorderTraversal.cpp
// problem: https://leetcode.com/problems/binary-tree-inorder-traversal/
// Runtime: 3 ms, faster than 60.90% of C++ online submissions for Binary Tree
// Inorder Traversal. Memory Usage: 8.5 MB, less than 39.59% of C++ online
// submissions for Binary Tree Inorder Traversal.
// Definition for a binary tree node.

#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
class Solution {
public:
  void dfs(TreeNode *node, vector<int> &ans) {
    if (!node)
      return;
    dfs(node->left, ans);
    ans.push_back(node->val);
    dfs(node->right, ans);
  }

  vector<int> inorderTraversal(TreeNode *root) {
    vector<int> ans;
    dfs(root, ans);
    return ans;
  }
};
